File: ruby/resistor-colors/resistor_colors.rb
module ResistorColors
  COLOR_VALUES = {
    'black' => 0,
    'brown' => 1,
    'red' => 2,
    'orange' => 3,
    'yellow' => 4,
    'green' => 5,
    'blue' => 6,
    'violet' => 7,
    'grey' => 8,
    'white' => 9
  }.freeze

  def self.value(colors)
    colors.map { |color| COLOR_VALUES[color] }.join.to_i
  end
end

# - Black: 0
# - Brown: 1
# - Red: 2
# - Orange: 3
# - Yellow: 4
# - Green: 5
# - Blue: 6
# - Violet: 7
# - Grey: 8
# - White: 9

File: haskell/resistor-color-trio/src/ResistorColors.hs
module ResistorColors
  ( Color(..)
  , Resistor(..)
  , label
  , ohms
  )
where


data Color =
    Black
  | Brown
  | Red
  | Orange
  | Yellow
  | Green
  | Blue
  | Violet
  | Grey
  | White
  deriving (Show, Enum, Bounded)

newtype Resistor = Resistor { bands :: (Color, Color, Color) }
  deriving Show

label :: Resistor -> String
label resistor = show display ++ " " ++ unit
 where
  (display, unit) = case ohms resistor of
    n | n == 0                  -> (n, "ohms")
    n | n `rem` 1000000000 == 0 -> (n `div` 1000000000, "gigaohms")
    n | n `rem` 1000000 == 0    -> (n `div` 1000000, "megaohms")
    n | n `rem` 1000 == 0       -> (n `div` 1000, "kiloohms")
    n                           -> (n, "ohms")


ohms :: Resistor -> Int
ohms (Resistor (a, b, c)) = (fromEnum a * 10 + fromEnum b) * 10 ^ fromEnum c
